pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertOffHeapMetadataManager.java

```diff
@@ -0,0 +1,253 @@
+  final ConcurrentHashMap<Integer, Object> _segmentIdToSegmentMap = new ConcurrentHashMap<>();
+  final AtomicInteger _segmentId = new AtomicInteger(INITIAL_SEGMENT_ID);
+
+  // TODO: figure out optimisation in lock usage. Might not be needed in some paths.
+  private final ReentrantReadWriteLock _readWriteLock = new ReentrantReadWriteLock();
+
+  public PartitionUpsertOffHeapMetadataManager(String tableNameWithType, int partitionId, ServerMetrics serverMetrics,
+      @Nullable PartialUpsertHandler partialUpsertHandler, HashFunction hashFunction)
+      throws Exception {
+    _tableNameWithType = tableNameWithType;
+    _partitionId = partitionId;
+    _serverMetrics = serverMetrics;
+    _partialUpsertHandler = partialUpsertHandler;
+    _hashFunction = hashFunction;
+
+    String prefix = Joiner.on("_").join("offHeapUpsert", _tableNameWithType, _partitionId);
+    Path offHeapStoreDir = Files.createTempDirectory(prefix);
+
+    LOGGER.info("Using offHeap storage dir {}", offHeapStoreDir);
+    String segmentName = StringUtil.join("-", "upsert_metadata", _tableNameWithType, String.valueOf(_partitionId),
+        String.valueOf(System.currentTimeMillis()));
+
+    // TODO: How to determine the cardinality of primary keys so as to estimate file size
+    _memoryManager =
+        new MmapMemoryManager(offHeapStoreDir.toAbsolutePath().toString(), segmentName,
+            null);
+
+    // Used to store mapping from primary key to an int id.
+    _bytesOffHeapMutableDictionary = new BytesOffHeapMutableDictionary(50000, 3, _memoryManager, null, 10);
+
+    //The record to be stored contains an integer segmentID (derived from segment), and integer doc id, and a long comparable.
+    // The comparable data type will need to be part of the constructor
+    // The offset at which the record is stored is determined by the dictId derived from the primary key using `_bytesOffHeapMutableDictionary`
+    _mutableForwardIndex =
+        new FixedByteSVMultiColForwardIndex(false, 10000, _memoryManager, null, new FieldSpec.DataType[]{
+            FieldSpec.DataType.INT, FieldSpec.DataType.INT, FieldSpec.DataType.LONG
+        });
+  }
+
+  /**
+   * Initializes the upsert metadata for the given immutable segment.
+   */
+  public void addSegment(IndexSegment segment, Iterator<RecordInfo> recordInfoIterator) {
+    String segmentName = segment.getSegmentName();
+    int segmentId = _segmentToSegmentIdMap.computeIfAbsent(segment, (segmentObj) -> {
+      Integer newSegmentId = _segmentId.incrementAndGet();
+      _segmentIdToSegmentMap.put(newSegmentId, segment);
+      return newSegmentId;
+    });
+    LOGGER.info("Adding upsert metadata for segment: {}", segmentName);
+
+    ThreadSafeMutableRoaringBitmap validDocIds = Objects.requireNonNull(segment.getValidDocIds());
+    while (recordInfoIterator.hasNext()) {
+      RecordInfo recordInfo = recordInfoIterator.next();
+      int primaryKeyId = getPrimaryKeyId(recordInfo);
```

Comment:
This is not thread safe. Multiple invocations can happen at the same time and mess up the dictionary